chain/update.go

```diff
@@ -0,0 +1,159 @@
+	}
+
+	HostUpdate struct {
+		Announcement chain.HostAnnouncement
+		BlockHeight  uint64
+		BlockID      types.BlockID
+		Timestamp    time.Time
+	}
+
+	WalletEventUpdate struct {
+		Addition bool
+		Event    wallet.Event
+	}
+
+	WalletOutputUpdate struct {
+		Addition bool
+		Element  wallet.SiacoinElement
+		ID       types.Hash256
+	}
+)
+
+// NewChainUpdate returns a new ChainUpdate.
+func NewChainUpdate(elements map[types.Hash256]types.StateElement) *Update {
+	return &Update{
+		StateElements:       elements,
+		ContractUpdates:     make(map[types.FileContractID]*ContractUpdate),
+		HostUpdates:         make(map[types.PublicKey]HostUpdate),
+		WalletOutputUpdates: make(map[types.Hash256]WalletOutputUpdate),
+	}
+}
+
+// ContractUpdate returns the ContractUpdate for the given file contract ID. If
+// it doesn't exist we initialize one with the given state.
+func (cu *Update) ContractUpdate(fcid types.FileContractID, state api.ContractState) *ContractUpdate {
+	_, ok := cu.ContractUpdates[fcid]
+	if !ok {
+		cu.ContractUpdates[fcid] = &ContractUpdate{State: state}
+	}
+	return cu.ContractUpdates[fcid]
+}
+
+// AddEvents is called with all relevant events added in the update.
+func (cu *Update) AddEvents(events []wallet.Event) error {
+	for _, event := range events {
+		cu.WalletEventUpdates = append(cu.WalletEventUpdates, WalletEventUpdate{
+			Addition: true,
+			Event:    event,
+		})
+	}
+	return nil
+}
+
+// AddSiacoinElements is called with all new siacoin elements in the
+// update. Ephemeral siacoin elements are not included.
+func (cu *Update) AddSiacoinElements(ses []wallet.SiacoinElement) error {
+	for _, se := range ses {
+		cu.WalletOutputUpdates[se.ID] = WalletOutputUpdate{
+			Addition: true,
+			ID:       se.ID,
+			Element:  se,
+		}
+	}
+	return nil
+}
+
+// RemoveSiacoinElements is called with all siacoin elements that were
+// spent in the update.
+func (cu *Update) RemoveSiacoinElements(ids []types.SiacoinOutputID) error {
+	for _, id := range ids {
+		cu.WalletOutputUpdates[types.Hash256(id)] = WalletOutputUpdate{
+			Addition: false,
+			ID:       types.Hash256(id),
+		}
+	}
+	return nil
+}
+
+// WalletStateElements returns all state elements in the database. It is used
+// to update the proofs of all state elements affected by the update.
+func (cu *Update) WalletStateElements() (elements []types.StateElement, _ error) {
+	// copy the output updates to the state
+	for id, el := range cu.WalletOutputUpdates {
+		cu.StateElements[id] = types.StateElement{
+			ID:          id,
+			LeafIndex:   el.Element.LeafIndex,
+			MerkleProof: el.Element.MerkleProof,
+		}
+	}
+	// loop over the state
+	for _, se := range cu.StateElements {
+		elements = append(elements, se)
+	}
+	return
+}
+
+// UpdateStateElements updates the proofs of all state elements affected by the
+// update.
+func (cu *Update) UpdateStateElements(elements []types.StateElement) error {
+	for _, se := range elements {
+		curr := cu.WalletOutputUpdates[se.ID]
```

Comment:
I think this suffers from the same issue as `WalletStateElements` and `RevertIndex`. 
It's not actually updating the whole database and only updates that we already have in `WalletOutputUpdates`.

I think the best course of action here is to drop the `Update` type. It is a middle step that we don't really need and only makes things more complicated and error prone. 

Instead we could move the `Update` to a interface that is implemented by the `SQLStore` directly and apply things without keeping anything in memory.
```
// TODO: combines wallet tx interfaces plus any methods we need to persist updates in the `SQLStore`. Implemented by SQLStore. No buffering needed. Writes directly to store.
type SubscriberTx interface {
  wallet.ApplyTx
  wallet.RevertTx

  UpdateContracts(...) error
  UpdateHosts(...) error
}

// TODO: apply all updates not applied by wallet.ApplyChainUpdates
func (cs *Subscriber) ApplyChainUpdates(tx SubscriberTx, ...) error {
  ...
}

// TODO: apply all updates not applied by wallet.RevertChainUpdates
func (cs *Subscriber) RevertChainUpdates(tx SubscriberTx, ...) error {
  ...
}

// TODO: handle syncing
func (cs *Subscriber) sync(index types.ChainIndex) error {
  // ...

  tx := sqlStore.BeginSubscriberTx() 

  wallet.RevertChainUpdates(tx, cs.WalletAddress, crus)
  cs.RevertChainUpdates(tx, crus)

  wallet.ApplyChainUpdates(tx, cs.WalletAddress, caus)
  cs.ApplyChainUpdates(tx, caus)

  tx.Commit()

  // ...
}

That way we can simply never end up in a position where we update in-memory state and not the database.

Reply:
I wanted to avoid ` tx := sqlStore.BeginSubscriberTx() ` at all costs but maybe keeping that state correct is too error prone indeed.